torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -247,6 +247,76 @@ static PyObject* applyBoolGetitem(const Variable& self, bool index) {
   }
 }
 
+static bool canDispatchToLegacyIndexing(const Variable& self, const variable_list& vars) {
+  // mask indexing
+  if (vars.size() == 1 && vars[0].type().scalarType() == ScalarType::Byte && vars[0].is_same_size(self)) {
+    return true;
+  }
+
+  // single tensor indexing
+  int num_defined_variables = 0;
+  for (auto& variable : vars) {
+    auto is_defined = variable.defined();
+    num_defined_variables += is_defined;
+    if (is_defined && (variable.dim() != 1 || variable.type().scalarType() != ScalarType::Long || variable.numel() == 0)) {
+      num_defined_variables = -1;
+      break;
+    }
+  }
+  if (num_defined_variables == 1) {
+    return true;
+  }
+  // advanced indexing
+  return false;
+}
+
+static Variable dispatch_legacy_index(const Variable& self, const variable_list& vars) {
+  if (vars.size() == 1 && vars[0].type().scalarType() == ScalarType::Byte && vars[0].is_same_size(self)) {
+    auto mask = vars[0];
+    if (self.type().backend() != mask.type().backend())
+      mask = mask.toBackend(self.type().backend());
```

Comment:
this will already be a no-op if they don't match, but I'm not even sure we want to do this -- what does indexing do exactly?  This could potentially try to convert sparse to non-sparse tensors.  Maybe just limit the scope of this for non-sparse?

Reply:
This was just meant to convert the tensors to be on the same device. If it could potentially convert to other layouts, then that was not intentional